src/core/ipe-setup.c

```diff
@@ -0,0 +1,127 @@
+                if (r == 0)
+                        break;
+
+                p = strstrafter(line, "policy_name=");
+                if (p && !policy_name) {
+                        policy_name = strdup(p);
+                        if (!policy_name)
+                                return log_oom();
+                        p = strchr(policy_name, ' ');
+                        if (p)
+                                *p = '\0';
+                }
+
+                p = strstrafter(line, "policy_version=");
+                if (p && !policy_version) {
+                        policy_version = strdup(p);
+                        if (!policy_version)
+                                return log_oom();
+                        p = strchr(policy_version, ' ');
+                        if (p)
+                                *p = '\0';
+                }
+
+                if (policy_name && policy_version)
+                        break;
+        }
+
+        if (!policy_name || !policy_version)
+                return log_error_errno(ENOENT, "Failed to find policy_name and policy_version in the IPE policy "IPE_POLICY_PATH);
+
+        policy_existing_version_file = strjoin(IPE_SECFS_POLICIES, policy_name, "/version");
+        if (!policy_existing_version_file)
+                return log_oom();
+
+        r = read_virtual_file(policy_existing_version_file, SIZE_MAX, &policy_existing_version, /* ret_size= */ NULL);
+        if (r < 0 && r != -ENOENT)
+                return log_error_errno(r, "Failed to read the existing IPE policy version: %m");
+        if (r >= 0) {
+                if (strverscmp_improved(policy_version, policy_existing_version) <= 0) {
+                        log_info("The existing IPE policy '%s' is at version '%s', the new version is '%s', skipping.",
+                                 policy_name,
+                                 policy_existing_version,
+                                 policy_version);
+                        return 0;
+                }
+
+                /* Policy is already installed and older, update it */
+                output_path_free = strjoin(IPE_SECFS_POLICIES, policy_name, "/update");
+                if (!output_path_free)
+                        return log_oom();
+                output_path = output_path_free;
+        } else {
+                /* Policy is not installed, install it and activate it */
+                output_path = IPE_SECFS_NEW_POLICY;
+        }
+
+        output = open(output_path, O_WRONLY|O_CLOEXEC);
+        if (output < 0)
+                return log_error_errno(errno, "Failed to open the IPE kernel interface %s, ignoring: %m", output_path);
+
+        /* The policy is inline signed in binary format, so it has to be copied in one go, otherwise the
+         * kernel will reject partial inputs with -EBADMSG. */
+        r = copy_bytes(input, output, UINT64_MAX, COPY_FSYNC);
```

Comment:
fsync, really?

Reply:
does this actually work? you are mixing stdio read and copy_bytes(), the read offset in the file should be undeterministic here

Reply:
if you want to mix stdio and raw reads, then maybe do an fd_reopen() in between to get a pristine fd back

Reply:
> does this actually work? you are mixing stdio read and copy_bytes(), the read offset in the file should be undeterministic here

Yes it does work - it's what fdopen_independent() already does, no? That's why I used it...

Not sure about fsync, it's sprinkled as snake oil, I can remove it

Reply:
ah, fdopen_independent() i didn't see. sorry for the confusion.

yes, please drop the fsync